### ISSUE TITLE
Test IATI datastore redirect and/or transition to new IATI datastore

### ISSUE THREAD

**Issue by matmaxgeds** [somaliaims/Somali-AIMSUI#582]
Hi Rashid, IATI are trying to setup a redirect from the datastore that we are currently using to the new one that they are implementing - this would mean that we did not have to change our parser.
I already submitted some comments from comparing just one activity here: https://docs.google.com/spreadsheets/d/1Iy4ZX1wYYXtG9uB1pMnI_I3mnanjPLRRSUEy2rbXo7Q/edit#gid=0
But we need to talk about how we could properly test it, and all the things that I am not aware of.

=========EMAIL FROM IATI=======
Please test datastore API redirects

We have now implemented these redirects on a test version of the old Datastore and as a user of the existing tool, we would welcome your input in this testing phase, as we are aware that your organisation makes (or has previously made) regular queries to the old Datastore.

To test the redirect functionality, use a regular API query and substitute “http://datastore.iatistandard.org” for “http://104.248.175.88”.

For example, to return all activities where Somalia is reported as a recipient-country, you could use this URL: http://datastore.iatistandard.org/api/1/access/activity.xml?recipient-country=SO&stream=true

To test the redirect, you should use the URL: http://104.248.175.88/api/1/access/activity.xml?recipient-country=SO&stream=true

Instead of immediately returning the results, you will be presented with the redirected query for the new Datastore. You should then run this query to check that the results correspond with your original regular query. Once the tool is fully launched, the redirected query will be run automatically, rather than just the URL being displayed.

It is important to note that we do not expect the number of activities that are returned to be the same when an equivalent query is made using the two different tools. The new Datastore does not import data that is published at version 1 of the IATI Standard or datasets that do not conform to the IATI Schema.

Rather, we would appreciate your feedback on whether the parameters translate correctly, and that the data that is returned corresponds with these parameters as expected. Please provide your feedback by Friday 15th June.

**Comment by raashidahmad** [comment]
Hi @matmaxgeds,
Thanks for all the research and be in co-ordination with IATI.
This is good to hear that they are introducing redirect. They have not implemented the redirect currently, when it is ready, we will be in a state to test it. For all the version 2 tasks, we will setup a testing server, and will test the features before pushing to prod.

**Comment by matmaxgeds** [comment]
@raashidahmad please use this for the redirect: https://iatidatastore.iatistandard.org/search/activity?q=recipient_country_code:(SO)&wt=xslt&tr=activity-xml.xsl&rows=5000000 they made a mess of automating the redirect

**Comment by matmaxgeds** [comment]
@raashidahmad we need to figure out how to test it e.g. maybe it is best to export just a few rows and have a look at it manually, or whether it is possible to write a script to pull the same IATI activity from other original source and the redirect and tell us how it differs?

**Comment by raashidahmad** [comment]
@matmaxgeds currently the redirect only returns the URL and nothing else. I am little confused about what i need to test. What i understood about the redirect so far is:
1. Currently to test the redirect, we will replace the provide IP with the address we are using for the query.
2. The redirect will return the query at the moment, so we need to grab the query from the redirect and then use that query to extract the desired results.
3. To test the redirect, i need to query this source as provided in the comment above which is, https://iatidatastore.iatistandard.org/search/activity?q=recipient_country_code:(SO)&wt=xslt&tr=activity-xml.xsl&rows=5000000

Is this correct? If it is, i think i will have to test it on my local system first with the current parser and see if it results okay. In any other case, i will have to study and compare both the results then. Please note we are going to query IATI just once per day at midnight.

**Comment by matmaxgeds** [comment]
@raashidahmad - yes, definitely local test only, and yes, use the source from 3. - they are still figuring out the redirect link. I guess the first thing to test is whether we will have to change anything in our parser e.g. to cope with the new header data etc. At the moment don't worry if there are different numbers of activities returned by the new link.

**Comment by raashidahmad** [comment]
@matmaxgeds Okay thanks. Will give it a try tonight.

**Comment by raashidahmad** [comment]
@matmaxgeds I have given a try to the URL given above. I set the IATI URL to above and results are below:

1. It extracted 0 organizations with the current parser
2. It extracted 0 sectors with the current parser
3. It extracted 0 activities with the current parser.
4. I manually looked at the parser and found that, there is a version change for IATI, it is 2.03, while the parser was written for 2.01. There is a difference with the data is being organized in the hierarchical manner.
5. I found some information is discarded, while some information is added in the new version of IATI.
6. Size of IATI file for the new version is 57MB while it is 64MB for the previous version.

In summary, we will write another parser for the newer version. There will not be many differences but changes in data's organization in the hierarchical manner. I am just making a rough guess we can write, test, and implement it within 7 days as there are many areas where we are using IATI, though keeping more days for it will be more practical guess to make it production ready.

**Comment by matmaxgeds** [comment]
@raashidahmad - thanks, let's discuss, as the previous datastore should also have been returning all the versions mixed together depending on the version that was used to publish each activity e.g. v1, v2.01, v.2.02 and v2.03, the new Datastore just converts all data to v2.03 - so this suggests that we might have had a problem before without knowing it?

Thanks for the time estimate - can you share an example of where the hierarchy issue means we have to refactor the parser so that I can understand it?

**Comment by raashidahmad** [comment]
@matmaxgeds sure, i will do some debugging in order to come up with some examples, so that we can understand and plan accordingly.

**Comment by raashidahmad** [comment]
@matmaxgeds During debugging i realized that because of the place of version match many of the tags were ignored by the parser and hence this was the reason that we got nothing with our current parser.

**Comment by matmaxgeds** [comment]
See more information here: https://discuss.iatistandard.org/t/iatis-datastore-and-validator-tools-launched/2021

**Comment by matmaxgeds** [comment]
As a user of the old datastore, I just wanted to flag that between now and 22 March 2021 users can choose to either use existing API calls to return data from the old Datastore, or use a Redirect version from the Old Datastore to the new Datastore. Redirects can be tested by replacing "datastore.iatistandard.org" in previous queries with "datastoreredirect.iatistandard.org." By using this version, existing API calls will be redirected and return data from the new Datastore. After 22 March 2021, the old Datastore will be deprecated and all queries to the Old Datastore will automatically be redirected to corresponding queries on the new Datastore.

You don’t need to edit the paths on any software using the old Datastore API. The redirects are temporarily at datastoreredirects.iatistandard.org, and will be replacing datastore.iatistandard.org once the 6 month period is up. All the documentation you are looking for is on the Datastore 2 API docs. If that is still not clear @Alex Miller can help to clarify this further.
